python/src/python-to-nix.cc

```diff
@@ -0,0 +1,164 @@
+        auto attrVal = pythonToNixValue(state, val);
+        if (!attrVal) {
+            return {};
+        }
+        attrs[state.symbols.create(name)] = attrVal;
+    }
+
+    return attrs;
+}
+
+nix::Value * pythonToNixValue(nix::EvalState & state, PyObject * obj)
+{
+    auto v = state.allocValue();
+
+    if (obj == Py_True && obj == Py_False) {
+        v->mkBool(obj == Py_True);
+    } else if (obj == Py_None) {
+        v->mkNull();
+    } else if (PyBytes_Check(obj)) {
+        auto str = checkNullByte(PyBytes_AS_STRING(obj), PyBytes_GET_SIZE(obj));
+        if (!str) {
+            return nullptr;
+        }
+
+        v->mkString(str);
+    } else if (PyUnicode_Check(obj)) {
+        Py_ssize_t size;
+        const char * utf8 = PyUnicode_AsUTF8AndSize(obj, &size);
+        auto str = checkNullByte(utf8, size);
+        if (!str) {
+            return nullptr;
+        }
+
+        v->mkString(utf8);
+    } else if (PyFloat_Check(obj)) {
+        v->mkFloat(PyFloat_AS_DOUBLE(obj));
+    } else if (PyLong_Check(obj)) {
+        v->mkInt(PyLong_AsLong(obj));
+    } else if (PyList_Check(obj)) {
+        v->mkList(PyList_GET_SIZE(obj));
+        for (Py_ssize_t i = 0; i < PyList_GET_SIZE(obj); i++) {
+            auto val = pythonToNixValue(state, PyList_GET_ITEM(obj, i));
+            if (!val) {
+                return nullptr;
+            }
+            v->listElems()[i] = val;
+        }
+    } else if (PyTuple_Check(obj)) {
+        v->mkList(PyTuple_GET_SIZE(obj));
+        for (Py_ssize_t i = 0; i < PyTuple_GET_SIZE(obj); i++) {
+            auto val = pythonToNixValue(state, PyTuple_GET_ITEM(obj, i));
+            if (!val) {
+                return nullptr;
+            }
+            v->listElems()[i] = val;
+        }
+    } else if (PyDict_Check(obj)) {
+        auto attrs = dictToAttrSet(obj, state);
+        if (!attrs) {
+            return nullptr;
+        }
+        auto attrsValue = attrs.value();
+        auto bindings = state.buildBindings(attrsValue.size());
+        for (auto & attr : attrsValue) {
+            bindings.insert(attr.first, attr.second);
+        }
+        v->mkAttrs(bindings);
+    }
+    return v;
+}
+
+std::optional<nix::StaticEnv> pythonToNixEnv(nix::EvalState & state, PyObject * vars, nix::Env ** env)
+{
+    Py_ssize_t pos = 0;
+    PyObject *key = nullptr, *val = nullptr;
+
+    *env = &state.allocEnv(vars ? PyDict_Size(vars) : 0);
+    (*env)->up = &state.baseEnv;
+
+    nix::StaticEnv staticEnv(false, state.staticBaseEnv.get());
+
+    if (!vars) {
+        return staticEnv;
+    }
+
+    auto displ = 0;
+    while (PyDict_Next(vars, &pos, &key, &val)) {
+        auto name = checkAttrKey(key);
+        if (!name) {
+            return {};
+        }
+
+        auto attrVal = pythonToNixValue(state, val);
+        if (!attrVal) {
+            return {};
+        }
+        staticEnv.vars.emplace_back(state.symbols.create(name), displ);
+        (*env)->values[displ++] = attrVal;
+    }
+
```

Comment:
```suggestion
staticEnv.sort()
```

I believe we should sort the env here, since python dict iteration happens in insertion order and nix wants alphabetical order.

Reply:
Good point, I also think we should probably replace staticEnv with just arguments passed to a function, so like `nix.eval("{ a, b }: a + b", args=dict('a': 1, 'b': 2))`